src/io/proxi.rs

```diff
@@ -1,12 +1,231 @@
-use std::fmt::Display;
-use std::str::FromStr;
+use std::{cmp::Ordering, fmt::Display, marker::PhantomData, str::FromStr};
 
+use num_traits::AsPrimitive;
 use serde::{Deserialize, Serialize};
 
-use super::usi::USI;
-use crate::params::{ControlledVocabulary, Param, ParamCow, Value, CURIE};
-use crate::spectrum::{ArrayType, BinaryArrayMap, BinaryDataArrayType, DataArray, IsolationWindowState, MultiLayerSpectrum, Precursor, ScanPolarity, SignalContinuity, SpectrumDescription};
-use crate::{curie, prelude::*};
+use crate::{
+    curie,
+    io::usi::USI,
+    params::{ControlledVocabulary, Param, ParamCow, Value, CURIE},
+    prelude::*,
+    spectrum::{
+        ArrayType, BinaryArrayMap, BinaryDataArrayType, DataArray, IsolationWindowState,
+        MultiLayerSpectrum, Precursor, ScanPolarity, SignalContinuity, SpectrumDescription,
+    },
+};
+
+/// The possible PROXI backends
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
+pub enum PROXIBackend {
+    PeptideAtlas,
+    MassIVE,
+    Pride,
+    Jpost,
+    ProteomeXchange,
+}
+
+impl PROXIBackend {
+    const ALL: &[Self] = &[
+        Self::PeptideAtlas,
+        Self::MassIVE,
+        Self::Pride,
+        Self::Jpost,
+        Self::ProteomeXchange,
+    ];
+
+    /// The PROXI server base url which needs concatenating of the USI at the end
+    const fn base_url(self) -> &'static str {
+        match self {
+            Self::PeptideAtlas => "http://www.peptideatlas.org/api/proxi/v0.1/spectra?resultType=full&usi=",
+            Self::MassIVE => "http://massive.ucsd.edu/ProteoSAFe/proxi/v0.1/spectra?resultType=full&usi=",
+            Self::Pride => "http://www.ebi.ac.uk/pride/proxi/archive/v0.1/spectra?resultType=full&usi=",
+            Self::Jpost => "https://repository.jpostdb.org/proxi/spectra?resultType=full&usi=",
+            Self::ProteomeXchange => "http://proteomecentral.proteomexchange.org/api/proxi/v0.1/spectra?resultType=full&usi=",
+        }
+    }
+}
+
+impl USI {
+    /// Retrieve this USI from the given PROXI backend. If no PROXI backend is indicated it will
+    /// aggregate the results from all known backends and return the first successful spectrum.
+    ///
+    /// This function is only available with the feature `proxi`.
+    pub fn get_spectrum_blocking(
+        &self,
+        backend: Option<PROXIBackend>,
+    ) -> Result<(PROXIBackend, Vec<PROXISpectrum>), PROXIError> {
+        backend.map_or_else(
+            || {
+                let client = reqwest::blocking::Client::new();
```

Comment:
It seems like creating a new `Client` on every request that the caller cannot configure might be wasteful or restrictive if the caller needs to use a network proxy. This implies there should be a separate `PROXIClient` type with blocking and async flavors that is bound to one (or many) `PROXIBackend`, and re-uses the HTTP client across requests.

This is a bunch of work though, so this idea can be deferred until someone actually wants it.

Reply:
Adding a parameter of `Option<Client>` works and is added now. This would allow the user to set up network proxies and the like.